src/reader.rs

```diff
@@ -138,52 +95,163 @@ impl<'a, R: Read> Reader<'a, R> {
         Ok(())
     }
 
+    fn fill_buf(&mut self, n: usize) -> Result<(), Error> {
+        // We don't have enough space in the buffer, need to grow it.
+        if n >= self.buf.capacity() {
+            self.buf.reserve(n);
+        }
+
+        unsafe { self.buf.set_len(n); }
+        self.reader.read_exact(&mut self.buf[..n])?;
+        self.buf_idx = 0;
+        Ok(())
+    }
+
     /// Try to read a data block, also performing schema resolution for the objects contained in
     /// the block. The objects are stored in an internal buffer to the `Reader`.
-    fn read_block(&mut self) -> Result<(), Error> {
-        match decode(&Schema::Long, &mut self.reader) {
-            Ok(block) => {
-                if let Value::Long(block_len) = block {
-                    if let Value::Long(block_bytes) = decode(&Schema::Long, &mut self.reader)? {
-                        let mut bytes = vec![0u8; block_bytes as usize];
-                        self.reader.read_exact(&mut bytes)?;
-
-                        let mut marker = [0u8; 16];
-                        self.reader.read_exact(&mut marker)?;
-
-                        if marker != self.marker {
-                            return Err(DecodeError::new(
-                                "block marker does not match header marker",
-                            ).into())
-                        }
-
-                        self.codec.decompress(&mut bytes)?;
-
-                        self.items.clear();
-                        self.items.reserve_exact(block_len as usize);
-
-                        let mut block_bytes = &bytes[..];
-                        for _ in 0..block_len {
-                            let item = from_avro_datum(
-                                &self.writer_schema,
-                                &mut block_bytes,
-                                self.reader_schema,
-                            )?;
-                            self.items.push_back(item)
-                        }
-
-                        return Ok(())
-                    }
+    fn read_block_next(&mut self) -> Result<(), Error> {
+        if !self.is_empty() {
+            // I suppose this isn't technically an error...
+            panic!("(TODO Replace this with an error) Invalid state; shouldn't call read_block_next");
```

Comment:
TODO left here :)

Reply:
Replaced with an `assert!`, this case represents a bug in the code, but won't necessarily prevent the `Reader` from continuing.